v5/core/iam_authenticator.go

```diff
@@ -249,7 +249,8 @@ func (authenticator *IamAuthenticator) setTokenData(tokenData *iamTokenData) {
 	// This was recommended by the IAM team to avoid problems in the future
 	// if the token service is changed to invalidate an existing refresh token
 	// when a new one is generated and returned in the response.
-	if tokenData != nil {
+	// Ignore if the refresh token is not_supported.
+	if tokenData != nil && tokenData.RefreshToken != "not_supported" {
```

Comment:
According to the IAM team, we should not be relying on "not_supported" to be the value when an invalid refresh token is returned by the IAM get_token operation.
Besides, this check wouldn't help the case where a valid refresh token is returned.
All you really need to do is to avoid calling Validate() and you will avoid the problem

Reply:
In that case, do you think we need to change Validate() logic somehow to accommodate the changes, maybe a private field called `refreshToken` which is used only within the module instead of overwriting user-supplied one?

I have just written a sample program for repro purpose, but this is very well possible if someone shares an authenticator object between multiple services then there is a possibility that the first service calls the Validate, Authenticate, and then the second service calls again Validate and Authenticate(the second one will fail because the first service already feed that refresh_token)